Possible double URL-decoding of form data

I tried to get SAML auth (pac4j) working with 2.x. It fails because the IDP response could not be parsed.

The response comes in a POST request's body, `application/x-www-form-urlencoded`. The value of the field of interest is a base64 encoded string, comes with proper URL encoding (i.e. `+` sign is encoded as %2B).

When pac4j tries to parse the response it will end up [here](https://github.com/pac4j/pac4j/blob/e6a3fbc762b02c7950b1a76457aa4f5c062ef034/pac4j-saml/src/main/java/org/pac4j/saml/transport/AbstractPac4jDecoder.java#L59), first trying to look for request parameters. This is implemented in `jooby-pac4j` [here](https://github.com/jooby-project/jooby/blob/dfaa1fa399463347a493737bbbdd93fe31d25e88/modules/jooby-pac4j/src/main/java/io/jooby/internal/pac4j/WebContextImpl.java#L40).

The parameter will be found in the `multipart` section. Jooby tries to populate a `FormData` object in `NettyContext#decodeForm(...)`. 

At [this line](https://github.com/jooby-project/jooby/blob/dfaa1fa399463347a493737bbbdd93fe31d25e88/modules/jooby-netty/src/main/java/io/jooby/internal/netty/NettyContext.java#L751), the value returned by `next.getString(UTF_8)` is already URL *decoded* by Netty.

Then shortly after [this line](https://github.com/jooby-project/jooby/blob/dfaa1fa399463347a493737bbbdd93fe31d25e88/jooby/src/main/java/io/jooby/internal/HashValue.java#L78) will be executed where it calls `decode(...)` on the value.

`FormdataNode` overrides `decode(...)` and calls `UrlParser.decode(value, false)` which will remove the `+` sign from the base64 encoded value, which will be invalid and therefore results to invalid XML as well which fails to parse.

I checked Jooby's test cases and I think (not completely sure though) that [this test case](https://github.com/jooby-project/jooby/blob/dfaa1fa399463347a493737bbbdd93fe31d25e88/tests/src/test/java/io/jooby/FeaturedTest.java#L3002) is invalid, because the `add(...)` method of `FormBody.Builder` is meant to add unencoded data, so I would expect to get the same string back. I'm guessing this because an `addEncoded(...)` method also exists. I found no docs for it unfortunately.

I tried to fix it by simply removing the override from `FormdataNode`.  Multipart requests also dont need it since there is no (url)encoding involved at all.

https://github.com/imeszaros/jooby/commit/8a3dd338b777a1efd19a909e92ca309867e9cd3e

I fixed the test case and added some new, results are interesting..

Only a single test fails, and only with Netty:

```
[ERROR] Failures: 
[ERROR]   FeaturedTest.rawVsDecodedValues:2989->lambda$rawVsDecodedValues$524:3003->lambda$null$520:3006 FeaturedTest.rawVsDecodedValues(Netty): expected: <a b/> but was: <a+b/>
```

Undetow and Jetty realize properly that [percent encoding](https://developer.mozilla.org/en-US/docs/Glossary/percent-encoding) used in `x-www-form-urlencoded` encodes spaces to `+`, but Netty doesn't.. :) It only decodes the `%__` sequence...

--- a/jooby/src/main/java/io/jooby/internal/FormdataNode.java
+++ b/jooby/src/main/java/io/jooby/internal/FormdataNode.java
@@ -12,8 +12,4 @@ public class FormdataNode extends HashValue implements Formdata {
   public FormdataNode(Context ctx) {
     super(ctx, null);
   }
-
-  @Override protected String decode(String value) {
-    return UrlParser.decode(value, false);
-  }
 }

--- a/jooby/src/main/java/io/jooby/internal/UrlParser.java
+++ b/jooby/src/main/java/io/jooby/internal/UrlParser.java
@@ -30,14 +30,10 @@ public final class UrlParser {
   }
 
   public static String decodePathSegment(String value) {
-    return decode(value, true);
-  }
-
-  public static String decode(String value, boolean isPath) {
     if (value == null || value.length() == 0) {
       return "";
     }
-    return decodeComponent(value, 0, value.length(), StandardCharsets.UTF_8, isPath);
+    return decodeComponent(value, 0, value.length(), StandardCharsets.UTF_8, true);
   }
 
   private static void decodeParams(HashValue root, String s, int from, Charset charset,

--- a/pom.xml
+++ b/pom.xml
@@ -68,7 +68,7 @@
     <!-- Server -->
     <undertow.version>2.1.3.Final</undertow.version>
     <jetty.version>9.4.31.v20200723</jetty.version>
-    <netty.version>4.1.50.Final</netty.version>
+    <netty.version>4.1.52.Final</netty.version>
     <boringssl.version>2.0.27.Final</boringssl.version>
 
     <!-- Reactive -->

--- a/tests/src/test/java/io/jooby/FeaturedTest.java
+++ b/tests/src/test/java/io/jooby/FeaturedTest.java
@@ -2999,15 +2999,32 @@ public class FeaturedTest {
         assertEquals("a b/", rsp.body().string().trim());
       });
 
+      // put in encoded value, expect urldecode(value)
       client.post("/form", new FormBody.Builder()
-          .add("v", "a+b%2f")
+          .addEncoded("v", "a+b%2f")
           .build(), rsp -> {
         assertEquals("a b/", rsp.body().string());
       });
 
+      // put in "raw" value, expect it as it is
+      client.post("/form", new FormBody.Builder()
+          .add("v", "a+b%2f")
+          .build(), rsp -> {
+        assertEquals("a+b%2f", rsp.body().string());
+      });
+
+      // no coding is done, expect it as it is
       client.post("/multipart", new MultipartBody.Builder()
           .setType(MultipartBody.FORM)
           .addFormDataPart("v", "a+b%2f")
+          .build(), rsp -> {
+        assertEquals("a+b%2f", rsp.body().string());
+      });
+
+      // no coding is done, expect it as it is
+      client.post("/multipart", new MultipartBody.Builder()
+          .setType(MultipartBody.FORM)
+          .addFormDataPart("v", "a b/")
           .build(), rsp -> {
         assertEquals("a b/", rsp.body().string());
       });

